extmod/modselect.c

```diff
@@ -26,62 +26,223 @@
  */
 
 #include "py/mpconfig.h"
-#if MICROPY_PY_SELECT
-
-#include <stdio.h>
-
 #include "py/runtime.h"
 #include "py/obj.h"
 #include "py/objlist.h"
 #include "py/stream.h"
 #include "py/mperrno.h"
 #include "py/mphal.h"
 
+#if MICROPY_PY_SELECT
+
+#if MICROPY_PY_SELECT_SELECT && MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+#error select.select is not supported with MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+#endif
+
+#if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+#include <poll.h>
+#endif
+
 // Flags for poll()
 #define FLAG_ONESHOT (1)
 
+// A single pollable object.
 typedef struct _poll_obj_t {
     mp_obj_t obj;
     mp_uint_t (*ioctl)(mp_obj_t obj, mp_uint_t request, uintptr_t arg, int *errcode);
+    #if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+    struct pollfd *pollfd;
```

Comment:
This needs a comment to explain that pollfd is mutually exclusive with flags and flags_ret. (I think that's what you're getting at with the `nonsys_` prefix, but it would be clearer to just spell it out and remove the prefix... although I guess the prefix stops you using the non-posix-opt code accidentally?).

Reply:
Also I think it would be good to be explicit that the pollfd member here points to the element in `poll_group_t::pollfds`.

Reply:
Comments added, and `nonsys_` prefix changed to `nonfd_`.

Also, `flags` renamed to `events` and `flags_ret` renamed to `revents` to make it more consistent and clearer (a `flags` argument is used elsewhere for completely unrelated things).